File: stepik-union/src/main/java/org/stepik/plugin/projectWizard/StepikProjectGenerator.java
package org.stepik.plugin.projectWizard;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ex.ProjectEx;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.exceptions.StepikClientException;
import org.stepik.api.objects.StudyObject;
import org.stepik.api.objects.courses.Course;
import org.stepik.core.StepikProjectManager;
import org.stepik.core.SupportedLanguages;
import org.stepik.core.courseFormat.StudyNode;
import org.stepik.core.courseFormat.StudyNodeFactory;
import org.stepik.core.metrics.Metrics;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static java.util.Collections.emptyList;
import static org.stepik.core.metrics.MetricsStatus.DATA_NOT_LOADED;
import static org.stepik.core.metrics.MetricsStatus.SUCCESSFUL;
import static org.stepik.core.metrics.MetricsStatus.TARGET_NOT_FOUND;
import static org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient;
import static org.stepik.core.stepik.StepikAuthManager.getCurrentUser;
import static org.stepik.plugin.projectWizard.CoursesList.COURSES;

public class StepikProjectGenerator {
    public static final StudyObject EMPTY_STUDY_OBJECT = initEmptyStudyNode();
    private static final Logger logger = Logger.getInstance(StepikProjectGenerator.class);
    private static StepikProjectGenerator instance;
    private static StudyNode projectRoot;
    private SupportedLanguages defaultLang = SupportedLanguages.INVALID;

    private StepikProjectGenerator() {
    }

    @NotNull
    private static StudyObject initEmptyStudyNode() {
        StudyObject course = new Course();
        course.setTitle("Empty");
        course.setDescription("Please, press refresh button");
        return course;
    }

    @NotNull
    public static StepikProjectGenerator getInstance() {
        if (instance == null) {
            instance = new StepikProjectGenerator();
        }
        return instance;
    }

    @NotNull
    public static CompletableFuture<List<StudyObject>> getCourses(@NotNull SupportedLanguages programmingLanguage) {
        return CompletableFuture.supplyAsync(() -> {
            List<Long> coursesIds = COURSES.getOrDefault(programmingLanguage, emptyList());

            if (coursesIds.isEmpty()) {
                return emptyList();
            }

            StepikApiClient stepikApiClient = authAndGetStepikApiClient();
            try {
                return stepikApiClient.courses()
                        .get()
                        .id(coursesIds)
                        .execute()
                        .getCourses()
                        .stream()
                        .map(course -> (StudyObject) course)
                        .sorted((course1, course2) -> {
                            long id1 = course1.getId();
                            long id2 = course2.getId();

                            int index1 = coursesIds.indexOf(id1);
                            int index2 = coursesIds.indexOf(id2);
                            return Integer.compare(index1, index2);
                        })
                        .collect(Collectors.toList());
            } catch (StepikClientException e) {
                logger.warn("Failed get courses", e);
                return emptyList();
            }
        });
    }

    public void createCourseNodeUnderProgress(@NotNull final Project project, @NotNull StudyObject data) {
        ProgressManager.getInstance()
                .runProcessWithProgressSynchronously(() -> {
                    if (data.getId() == 0) {
                        logger.warn("Failed to get a course");
                        Metrics.createProject(project, DATA_NOT_LOADED);
                        return;
                    }

                    ProgressIndicator indicator = ProgressManager.getInstance().getProgressIndicator();
                    indicator.setIndeterminate(true);

                    StepikApiClient stepikApiClient = authAndGetStepikApiClient();
                    projectRoot = StudyNodeFactory.createTree(project, stepikApiClient, data);
                }, "Creating Project", true, project);
    }

    public void generateProject(@NotNull Project project) {
        StepikProjectManager stepikProjectManager = StepikProjectManager.getInstance(project);
        if (stepikProjectManager == null) {
            Metrics.createProject(project, TARGET_NOT_FOUND);
            return;
        }
        stepikProjectManager.setRootNode(projectRoot);
        stepikProjectManager.setDefaultLang(getDefaultLang());
        stepikProjectManager.setCreatedBy(getCurrentUser().getId());

        ((ProjectEx) project).setProjectName(projectRoot.getName());

        Metrics.createProject(project, SUCCESSFUL);
    }

    @NotNull
    public SupportedLanguages getDefaultLang() {
        return defaultLang;
    }

    public void setDefaultLang(@NotNull SupportedLanguages defaultLang) {
        this.defaultLang = defaultLang;
    }
}

File: stepik-union/src/main/kotlin/org/stepik/plugin/projectWizard/StepikProjectGenerator.kt
package org.stepik.plugin.projectWizard

import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.ex.ProjectEx
import org.stepik.api.exceptions.StepikClientException
import org.stepik.api.objects.StudyObject
import org.stepik.core.ProjectGenerator
import org.stepik.core.StudyUtils.getProjectManager
import org.stepik.core.SupportedLanguages
import org.stepik.core.common.Loggable
import org.stepik.core.courseFormat.StudyNode
import org.stepik.core.courseFormat.StudyNodeFactory
import org.stepik.core.metrics.Metrics
import org.stepik.core.metrics.MetricsStatus.DATA_NOT_LOADED
import org.stepik.core.metrics.MetricsStatus.TARGET_NOT_FOUND
import org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient
import org.stepik.core.stepik.StepikAuthManager.currentUser
import org.stepik.plugin.projectWizard.CoursesList.COURSES
import java.util.Collections.emptyList
import java.util.concurrent.CompletableFuture

object StepikProjectGenerator : ProjectGenerator, Loggable {
    var defaultLang = SupportedLanguages.INVALID

    fun createCourseNodeUnderProgress(project: Project, data: StudyObject) {
        ProgressManager.getInstance()
                .runProcessWithProgressSynchronously({
                    if (data.id == 0L) {
                        logger.warn("Failed to get a course")
                        Metrics.createProject(project, DATA_NOT_LOADED)
                        return@runProcessWithProgressSynchronously
                    }

                    val indicator = ProgressManager.getInstance().progressIndicator
                    indicator.isIndeterminate = true

                    val stepikApiClient = authAndGetStepikApiClient()
                    projectRoot = StudyNodeFactory.createTree(project, stepikApiClient, data)
                }, "Creating Project", true, project)
    }

    fun generateProject(project: Project) {
        val projectManager = getProjectManager(project)
        if (projectManager == null) {
            Metrics.createProject(project, TARGET_NOT_FOUND)
            return
        }
        projectManager.projectRoot = projectRoot
        projectManager.createdBy = currentUser.id
        projectManager.defaultLang = defaultLang

        (project as ProjectEx).setProjectName(projectRoot!!.name)

        Metrics.createProject(project)
    }


    private var projectRoot: StudyNode? = null


    fun getCourses(programmingLanguage: SupportedLanguages): CompletableFuture<List<StudyObject>> {
        return CompletableFuture.supplyAsync {
            val coursesIds = COURSES.getOrDefault(programmingLanguage, emptyList())

            if (coursesIds.isEmpty()) {
                return@supplyAsync emptyList<StudyObject>()
            }

            val stepikApiClient = authAndGetStepikApiClient()

            try {
                return@supplyAsync stepikApiClient.courses()
                        .get()
                        .id(coursesIds)
                        .execute()
                        .courses
                        .map { course -> course }
                        .sortedBy { coursesIds.indexOf(it.id) }
            } catch (e: StepikClientException) {
                logger.warn("Failed get courses", e)
                return@supplyAsync emptyList<StudyObject>()
            }
        }
    }
}
